crypto/src/asset/amount.rs

```diff
@@ -1,26 +1,54 @@
 use crate::{Fq, Fr};
+use anyhow::anyhow;
 use penumbra_proto::{core::crypto::v1alpha1 as pb, Protobuf};
 use serde::{Deserialize, Serialize};
-use std::{fmt::Display, num::NonZeroU64, ops};
+use std::{fmt::Display, num::NonZeroU128, ops};
 
 #[derive(Serialize, Deserialize, PartialEq, PartialOrd, Eq, Clone, Debug, Copy)]
 #[serde(try_from = "pb::Amount", into = "pb::Amount")]
 pub struct Amount {
-    // TODO(ava): update to u128
-    pub inner: u64,
+    inner: u128,
+}
+
+impl Amount {
+    pub fn value(&self) -> u64 {
+        self.inner as u64
+    }
 }
 
 impl From<Amount> for pb::Amount {
     fn from(a: Amount) -> Self {
-        pb::Amount { lo: a.inner, hi: 0 }
+        let value = a.inner;
+        if value <= u64::MAX as u128 {
+            pb::Amount {
+                lo: value.try_into().unwrap(),
+                hi: 0,
+            }
+        } else {
+            let bytes = value.to_le_bytes();
+            let mut lo_bytes: [u8; 8] = [0; 8];
+            let mut hi_bytes: [u8; 8] = [0; 8];
+
+            for i in 0..16 {
+                if i < 8 {
+                    lo_bytes[i] = bytes[i];
+                } else {
+                    hi_bytes[i % 8] = bytes[i];
+                }
+            }
+
+            let lo = u64::from_le_bytes(lo_bytes);
+            let hi = u64::from_le_bytes(hi_bytes);
+            pb::Amount { lo, hi }
+        }
```

Comment:
This seems a bit complicated. I think it would be simpler to do something like


let lo = value as u64;
let hi = (value >> 64) as u64;

to get the high and low 64 bit parts directly.

Reply:
oh, this is nice!